framework/DataObjects/DataSet.py

```diff
@@ -0,0 +1,1853 @@
+        else:
+          index,rlz = self._getRealizationFromCollectorByValue(matchDict,tol=tol)
+      # otherwise, first try to find it in the data
+      else:
+        index,rlz = self._getRealizationFromDataByValue(matchDict,tol=tol, unpackXArray=unpackXArray)
+        # if no match found in data, try in the collector (if there's anything in it)
+        if rlz is None:
+          if numInCollector > 0:
+            index,rlz = self._getRealizationFromCollectorByValue(matchDict,tol=tol)
+      return index,rlz
+
+  def remove(self,realization=None,variable=None):
+    """
+      Used to remove either a realization or a variable from this data object.
+      @ In, realization, dict or int, optional, (matching or index of) realization to remove
+      @ In, variable, str, optional, name of "column" to remove
+      @ Out, None
+    """
+    if self.size == 0:
+      self.raiseAWarning('Called "remove" on DataObject, but it is empty!')
+      return
+    noData = self._data is None or len(self._data) == 0
+    noColl = self._collector is None or len(self._collector) == 0
+    assert(not (realization is None and variable is None))
+    assert(not (realization is not None and variable is not None))
+    assert(self._data is not None)
+    # TODO what about removing from collector?
+    if realization is not None:
+      # TODO reset scaling factors
+      self.raiseAnError(NotImplementedError,'TODO implementation for removing realizations is not currently implemented!')
+    elif variable is not None:
+      # remove from self._data
+      if not noData:
+        self._data = self._data.drop(variable)
+      # remove from self._collector
+      if not noColl:
+        varIndex = self._orderedVars.index(variable)
+        self._collector.removeEntity(varIndex)
+      # remove references to variable in lists
+      self._orderedVars.remove(variable)
+      # TODO potentially slow lookups
+      for varlist in [self._inputs,self._outputs,self._metavars]:
+        if variable in varlist:
+          varlist.remove(variable)
+      # remove from pivotParams, and remove any indexes without keys
+      for pivot in self.indexes:
+        if variable in self._pivotParams[pivot]:
+          self._pivotParams[pivot].remove(variable)
+        if len(self._pivotParams[pivot]) == 0:
+          del self._pivotParams[pivot]
+          # if in self._data, clear the index
+          if not noData and pivot in self._data.dims:
+            del self._data[pivot] # = self._data.drop(pivot,dim
+      # TODO remove references from general metadata?
+    if self._scaleFactors is not None:
+      self._scaleFactors.pop(variable,None)
+    #either way reset kdtree
+    self.inputKDTree = None
+
+  def renameVariable(self,old,new):
+    """
+      Changes the name of a variable from "old" to "new".
+      @ In, old, str, old name
+      @ In, new, str, new name
+      @ Out, None
+    """
+    # determine where the old variable was
```

Comment:
Are all of these necessary, or do some preclude others?

Reply:
Finding if it's in the list takes the same amount of time as getting the index. Getting the index would eliminate iterating over the list multiple times.

Reply:
Github seems to have misplaced these comments, as they don't seem to fit this line.

Reply:
The future-proofing attempt here was if the variable could be found in multiple places, perhaps even multiple times in a place.  Since renaming is not a frequent operation, it should not be a bottleneck.